pkg/nsx/services/builder.go

```diff
@@ -0,0 +1,1351 @@
+package services
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/data"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+func (service *SecurityPolicyService) buildSecurityPolicy(obj *v1alpha1.SecurityPolicy) (*model.SecurityPolicy,
+	*[]model.Group, error) {
+	var nsxRules []model.Rule
+	var nsxGroups []model.Group
+	log.V(1).Info("building the model SecurityPolicy from CR SecurityPolicy", "object", *obj)
+	nsxSecurityPolicy := &model.SecurityPolicy{}
+
+	policyID := fmt.Sprintf("sp_%s", obj.UID)
+	nsxSecurityPolicy.Id = &policyID
+
+	policyName := fmt.Sprintf("%s-%s", obj.ObjectMeta.Namespace, obj.ObjectMeta.Name)
+	nsxSecurityPolicy.DisplayName = &policyName
+
+	// TODO: confirm the sequence number: offset
+	policyPriority := int64(obj.Spec.Priority)
+	nsxSecurityPolicy.SequenceNumber = &policyPriority
+
+	policyGroup, policyGroupPath, err := service.buildPolicyGroup(obj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	nsxSecurityPolicy.Scope = []string{policyGroupPath}
+	if policyGroup != nil {
+		nsxGroups = append(nsxGroups, *policyGroup)
+	}
+
+	for ruleIdx, rule := range obj.Spec.Rules {
+		// A rule containing named port may expand to multiple rules if the name maps to multiple port numbers.
+		nsxRule, ruleGroups, err := service.buildRuleAndGroups(obj, &rule, ruleIdx)
+		if err != nil {
+			return nil, nil, err
```

Comment:
Don't remove `log.Error(err, "failed to build rule and groups", "rule", rule, "ruleIndex", i)` before return the error, unless we have another way to know the details of the rule